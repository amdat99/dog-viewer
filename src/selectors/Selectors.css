.selector {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  margin-left: 10px;
  width: 150px;
}
.selector-num {
  margin-left: 10px;
  width: 150px;
}

.form-control {
  margin-top: 5px;
  position: relative;
}

.form-control2 {
  margin-top: 5px;
  position: relative;
}

@media screen and (max-width: 800px) {
  .img-num {
    width: 140px;
  }

  .selector-num {
    position: absolute;
    margin-top: 20px;
    top: 10%;
    left: 5%;
  }
}
