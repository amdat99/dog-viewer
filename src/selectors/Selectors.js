import React,{useState, useEffect } from 'react'
import {fetchData} from '../utils'



 function Selectors({currentBreed,setCurrentBreed ,currentSubBreedList , setCurrentImgNum, currentImgNum,
    setCurrentSubBreedList, breeds, currentSubBreed,setCurrentSubBreed, numOfImgs,getImageLength   }) {
    
    const [hasSubBreed, setHasSubBreed] = useState(true)
    
 

    useEffect(() =>{
        onChangeBreedSelect()
        if(currentBreed){
        fetchData(`https://dog.ceo/api/breed/${currentBreed}/list`).then(data => {
          
        if(!data.message.length ){
            setHasSubBreed(false)
            setCurrentSubBreed('')
            
          } else{
            setCurrentSubBreedList(data.message)
            setCurrentSubBreed(data.message[0])
          }
       
        })
    }   
    //eslint-disable-next-line
      },[currentBreed, setCurrentBreed]
      )
    
      useEffect(()=>{
        getImageLength()
      },[setCurrentSubBreed,currentSubBreed,getImageLength])


      const onChangeBreedSelect = () =>{
        setHasSubBreed(true)
        setCurrentSubBreedList(null)
        setCurrentSubBreed('')
        getImageLength()
      }


    return (
    <React.Fragment>
        <div className="selector">
          
          <label>Breed</label>
            {breeds?
              <select className={"form-control"}
              value={currentBreed}
              onChange={(e)=>setCurrentBreed(e.target.value)}
              >
              {breeds.map((breed, i) => {
                  return  <option key={i} value={breed}>{breed}</option>
              })
              }
            </select>
          :null }
          </div>

          <div className="selector">
          <label>Sub Breed</label>
            
              <select id={hasSubBreed ? null : "error-border" }   className={"form-control"}
              value={currentSubBreed}
              onChange={(e)=>setCurrentSubBreed(e.target.value)}
              >
            {currentSubBreedList?
            currentSubBreedList.map((breed, i) => {
                return  <option 
                key={i} value={breed}>{breed}</option>
            })
            :null}
           
            </select>
            {currentSubBreed.length > 1?
             <span id='sub-reset' onClick={()=>setCurrentSubBreed('')}> all sub breeds</span>
            :null}
        </div>

        <div className="selector">
            <label>Num of Images ({numOfImgs})</label>
             <input id ={currentImgNum > numOfImgs ? 'error-border' : null}
             type="number" onChange={(e)=>setCurrentImgNum(e.target.value)}  name="quantity" min="1" max={numOfImgs} />
        </div>
        
        <button id='view-button'  type="submit" >View Images</button>
    </React.Fragment>

       )
}

export default Selectors