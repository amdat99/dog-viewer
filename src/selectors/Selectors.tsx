import React, { useState, useEffect } from "react";
import { fetchData } from "../utils";

import "./Selectors.css";

type selectorProps = {
  currentBreed: string;
  setCurrentBreed: (breed: string) => void;
  currentSubBreedList: Array<string>;
  breeds: Array<string>;
  currentSubBreed: string;
  currentImgNum: number;
  setCurrentImgNum: (value: number) => void;
  numOfImgs: number;
  setCurrentSubBreed: (value: string) => void;
  getImageLength: () => void;
  setCurrentSubBreedList: (value: string[]) => void;
};

function Selectors({
  currentBreed,
  setCurrentBreed,
  currentSubBreedList,
  setCurrentImgNum,
  currentImgNum,
  setCurrentSubBreedList,
  breeds,
  currentSubBreed,
  setCurrentSubBreed,
  numOfImgs,
  getImageLength,
}: selectorProps) {
  const [hasSubBreed, setHasSubBreed] = useState(true);

  const onChangeBreedSelect = () => {
    setHasSubBreed(true);
    setCurrentSubBreedList([]);
    setCurrentSubBreed("");
  };

  useEffect(() => {
    onChangeBreedSelect();
    if (currentBreed) {
      fetchData(`https://dog.ceo/api/breed/${currentBreed}/list`).then(
        (data) => {
          if (!data.message.length) {
            setHasSubBreed(false);
            setCurrentSubBreed("");
          } else {
            setCurrentSubBreedList(data.message);
            setCurrentSubBreed(data.message[0]);
          }
        }
      );
    }
    //eslint-disable-next-line
  }, [currentBreed, setCurrentBreed]);

  useEffect(() => {
  getImageLength()
  }, [setCurrentSubBreed, currentSubBreed, getImageLength]);

  return (
    <React.Fragment>
      <div className="selector">
        <label>Breed</label>

        <select
          id="breed-select"
          className={"form-control"}
          value={currentBreed}
          onChange={(e) => setCurrentBreed(e.target.value)}
        >
          {breeds.map((breed, i) => {
            return (
              <option id="breed-option" key={i} value={breed}>
                {breed}
              </option>
            );
          })}
        </select>
      </div>

      <div className="selector">
        <label>Sub Breed</label>

        <select
          // @ts-ignore */
          id={hasSubBreed ? null : "error-border"}
          className={"form-control2"}
          value={currentSubBreed}
          onChange={(e) => setCurrentSubBreed(e.target.value)}
        >
          {currentSubBreedList
            ? currentSubBreedList.map((breed, i) => {
                return (
                  <option key={i} value={breed}>
                    {breed}
                  </option>
                );
              })
            : null}
        </select>
        {currentSubBreed && currentSubBreed.length > 1 ? (
          <span id="sub-reset" onClick={() => setCurrentSubBreed("")}>
            {" "}
            all sub breeds
          </span>
        ) : null}
      </div>

      <div className="selector-num">
        <label>Num of Images ({numOfImgs})</label>
        <input
          // @ts-ignore */
          id={currentImgNum > numOfImgs ? "error-border" : null}
          type="number"
          className="img-num"
          onChange={(e: React.ChangeEvent<HTMLInputElement>): void => {
            setCurrentImgNum(parseInt(e.target.value));
          }}
          name="quantity"
          min="1"
          max={numOfImgs}
        />
      </div>
    </React.Fragment>
  );
}

export default Selectors;
