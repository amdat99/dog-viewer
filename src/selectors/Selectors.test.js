import React from "react";

import { shallow,mount } from "enzyme";
import { createSerializer } from "enzyme-to-json";
import toJson from "enzyme-to-json";

import  Selectors  from "./Selectors";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));


let mockFetchImages = jest.fn();

let mockSetCurrentBreed = jest.fn();
let mockSetCurrentSubBreed = jest.fn();
let mockSetCurrentImgNum = jest.fn();

let wrapper;
let wrapper2
let wrapper3;

beforeEach(() => {
  const mockProps = {
    message:[
      "https://images.dog.ceo/breeds/ovcharka-caucasian/IMG_20190528_194200.jpg",
      "https://images.dog.ceo/breeds/ovcharka-caucasian/IMG_20190601_185700.jpg",
      
    ],

    fetchImages: mockFetchImages,
    setCurrentBreed: mockSetCurrentBreed,
    setCurrentSubBreed: mockSetCurrentSubBreed,
    setCurrentImgNum : mockSetCurrentImgNum,
  };

  wrapper = toJson(shallow(<Selectors {...mockProps} />));
  wrapper2 = shallow(<Selectors {...mockProps} />);
  
});

it("expect to render selectors component", () => {
  expect(wrapper2).toMatchSnapshot();
  console.log(wrapper2.debug());

});


it("expect setCurrentBreed to be called on onChange", () => {
    
    let breedSelect = wrapper2.find("select").first()
    breedSelect.simulate("change", { target: { value: "akita" } });
    breedSelect = wrapper2.find("select").first()
    expect(mockSetCurrentBreed).toHaveBeenCalled();

});

    it("expect setCurrentSubBreed to be called on onChange", () => {

        
        let subBreedSelect = wrapper2.find(".form-control2");
        subBreedSelect.simulate("change", { target: { value: ["english","american"] } });
        subBreedSelect = wrapper2.find(".form-control2");
        expect(mockSetCurrentSubBreed).toHaveBeenCalled();
         });

    
 

    it("expect SetCurrentImgNum to be called on onChange", () => {

         
            let subBreedSelect = wrapper2.find(".img-num");
            subBreedSelect.simulate("change", { target: { value: 1 } });
            subBreedSelect = wrapper2.find(".img-num");
            expect(mockSetCurrentImgNum).toHaveBeenCalled();
            
            });