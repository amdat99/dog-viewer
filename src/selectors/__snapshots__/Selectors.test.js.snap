// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render selectors component 1`] = `
<Fragment>
  <div
    className="selector"
  >
    <label>
      Breed
    </label>
    <select
      className="form-control"
      id="breed-select"
      onChange={[Function]}
    />
    )
  </div>
  <div
    className="selector"
  >
    <label>
      Sub Breed
    </label>
    <select
      className="form-control2"
      id={null}
      onChange={[Function]}
    />
  </div>
  <div
    className="selector"
  >
    <label>
      Num of Images (
      )
    </label>
    <input
      className="img-num"
      id={null}
      min="1"
      name="quantity"
      onChange={[Function]}
      type="number"
    />
  </div>
</Fragment>
`;
